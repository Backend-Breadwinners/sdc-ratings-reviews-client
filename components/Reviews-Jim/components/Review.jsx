import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Container, Row, Col } from 'react-bootstrap';
import Rating from '@material-ui/lab/Rating';

const Review = ({ review, markHelpful }) => {
  const [helpful, setHelpful] = useState(false);
  const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
  const dateVal = new Date(review.date);
  const month = monthNames[dateVal.getMonth()];
  const day = dateVal.getDate();
  const year = dateVal.getFullYear();

  return (
    <Container className="review-card">
      <Row>
        <Col>
          <Rating className="star-rating" value={review.rating} precision={0.25} readOnly />
        </Col>
        <Col>
          {`${review.reviewer_name}, ${month} ${day}, ${year}`}
        </Col>
      </Row>
      <Row>
        <Col>
          {review.summary}
        </Col>
      </Row>
      <Row>
        <Col>
          {review.body}
        </Col>
      </Row>
      <Row>
        {review.photos.map((photo) => <img key={photo.id} alt="" src={photo.url} />)}
      </Row>
      {review.recommended ? <Row><Col>âœ“ I recommend this product</Col></Row> : null}
      {review.response ? (
        <Row>
          <Col>
            Response:
            {' '}
            {review.response}
          </Col>
        </Row>
      ) : null}
      <Row>
        {helpful ? (
          <Col>
            Marked as helpful! (
            {review.helpfulness}
            )
          </Col>
        ) : (
          <Col>
            Was this review helpful?
            {' '}
            <span onClick={(e) => { markHelpful(e, review.review_id); setHelpful(true); }}><u>Yes</u></span>
            {' '}
            (
            {review.helpfulness}
            ) |
            {' '}
            <u>Report</u>
          </Col>
        )}
      </Row>

      <style jsx>
        {`
          img {
            height: 150px;
          }
      `}
      </style>

    </Container>
  );
};

Review.propTypes = {
  review: PropTypes.shape({
    rating: PropTypes.number.isRequired,
    reviewer_name: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    summary: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired,
    photos: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      url: PropTypes.string.isRequired,
    })),
    recommended: PropTypes.bool,
    response: PropTypes.string,
    helpfulness: PropTypes.number.isRequired,
  }),
  markHelpful: PropTypes.func.isRequired,
};

Review.defaultProps = {
  review: null,
};

export default Review;
