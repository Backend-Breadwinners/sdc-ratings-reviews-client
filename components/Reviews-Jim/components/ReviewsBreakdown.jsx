import React, { useState, useEffect } from 'react';
import axios from 'axios';
import ProductBreakdown from './ProductBreakdown';
import RatingBreakdown from './RatingBreakdown';
import Stars from './Stars';
import config from '../../../config';

const ReviewsBreakdown = ({ productId }) => {
  const [productMeta, setProductMeta] = useState(null);
  const [rating, setRating] = useState(null);
  const [recommended, setRecommended] = useState(null);

  const getProductMeta = (product) => {
    const api = 'https://app-hrsei-api.herokuapp.com/api/fec2/hr-bld/reviews/meta';
    const options = {
      url: `${api}?product_id=${product}`,
      method: 'get',
      headers: {
        Authorization: config.TOKEN,
      },
    };

    axios(options)
      .then((res) => {
        setProductMeta(res.data);
      })
      .catch((err) => { console.log(err); });
  };

  const ratingCreator = (ratingsObj) => {
    let allRatings = 0;
    let ratingCount = 0;
    for (const key in ratingsObj) {
      for (let i = 0; i < ratingsObj[key]; i++) {
        allRatings += Number(key);
        ratingCount++;
      }
    }
    return setRating((allRatings / ratingCount).toFixed(1));
  };

  const reviewPercentage = (ratings, recommended) => {
    let totalRatings = 0;
    for (const key in ratings) {
      totalRatings += Number(ratings[key]);
    }
    return setRecommended(Math.floor((Number(recommended) / totalRatings) * 100));
  };

  useEffect(() => {
    getProductMeta(productId);
  }, []);

  useEffect(() => {
    if (productMeta) {
      ratingCreator(productMeta.ratings);
      reviewPercentage(productMeta.ratings, productMeta.recommended.true);
    }
  }, [productMeta]);

  return (
    <div>
      <div><h4>{rating}</h4></div>
      <Stars rating={rating} />
      <div>
        {recommended}
        % of reviews recommend this product
      </div>
      <RatingBreakdown />
      <ProductBreakdown />
    </div>
  );
};

export default ReviewsBreakdown;
