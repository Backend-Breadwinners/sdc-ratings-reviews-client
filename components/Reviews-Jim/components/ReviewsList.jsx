/* eslint-disable */

import React, { useState, useEffect } from 'react';
import axios from 'axios';
import config from '../../../config';
import Review from './Review';

const ReviewsList = ({ productId }) => {
  const [ reviews, setReviews ] = useState([]);
  const [ reviewCount, setReviewCount ] = useState(2);
  const [ sorted, setSorted ] = useState('relevant');

  const getProductReviews = (product, count, sorted) => {
    const api = 'https://app-hrsei-api.herokuapp.com/api/fec2/hr-bld/reviews/';
    const options = {
      url: `${api}?product_id=${product}&count=${count}&sort=${sorted}`,
      method: 'get',
      headers: {
        Authorization: config.TOKEN,
      },
    };

    axios(options)
      .then((res) => {
        setReviews(res.data.results);
      })
      .catch((err) => { console.log(err); });
  };

  const addTwoReviews = (e) => {
    e.preventDefault();
    setReviewCount(reviewCount + 2);
  }

  const handleSortedChange = (e) => {
    e.preventDefault();
    setSorted(e.target.value);
  }

  useEffect(() => {
    getProductReviews(productId, reviewCount, sorted);
  }, [reviewCount, sorted]);

  return (
    <div>
      <div>XXX reviews, sorted by
        <select value={sorted} onChange={(e) => handleSortedChange(e)}>
          <option value="relevant">Relevant</option>
          <option value="helpful">Helpful</option>
          <option value="newest">Newest</option>
        </select>
      </div>
      {reviews.map((review) => <Review key={review.review_id} review={review} />)}
      <button onClick={(e) => addTwoReviews(e)}>More Reviews</button>
      <button>Add a Review +</button>
    </div>
  );
};

export default ReviewsList;