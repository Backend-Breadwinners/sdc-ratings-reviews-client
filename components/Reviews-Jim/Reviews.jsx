/* eslint-disable no-console */
import React, { useState, useEffect } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import axios from 'axios';
import config from '../../config';
import 'bootstrap/dist/css/bootstrap.min.css';

import ReviewsList from './components/ReviewsList';
import ReviewsBreakdown from './components/ReviewsBreakdown';

const Reviews = ({
  productId, setProductRating, reviewsRef, productName,
}) => {
  const [productReviews, setProductReviews] = useState(null);
  const [productMeta, setProductMeta] = useState(null);
  const [sortStatus, setSortStatus] = useState('relevant');
  const [renderToggle, setRenderToggle] = useState(false);
  const [selectedRatings, setSelectedRatings] = useState([]);
  const [ratingsLength, setRatingsLength] = useState(selectedRatings.length);

  const getProductReviews = (product, sort, count = null) => {
    let api = `https://app-hrsei-api.herokuapp.com/api/fec2/hr-bld/reviews/?product_id=${product}&sort=${sort}`;

    if (count) {
      api += `&count=${count}`;
    }

    const options = {
      url: api,
      method: 'get',
      headers: {
        Authorization: config.TOKEN,
      },
    };

    axios(options)
      .then((res) => {
        setProductReviews(res.data.results);
      })
      .then(() => {
        setRenderToggle(true);
      })
      .catch((err) => { console.log('GET REVIEWS ERROR ', err); });
  };

  const getProductMeta = (product) => {
    const api = 'https://app-hrsei-api.herokuapp.com/api/fec2/hr-bld/reviews/meta';
    const options = {
      url: `${api}?product_id=${product}`,
      method: 'get',
      headers: {
        Authorization: config.TOKEN,
      },
    };

    axios(options)
      .then((res) => {
        setProductMeta(res.data);
      })
      .catch((err) => { console.log('GET META ERROR ', err); });
  };

  const handleSortChange = (e) => {
    e.preventDefault();
    setSortStatus(e.target.value);
  };

  useEffect(() => {
    getProductReviews(productId, sortStatus);
    getProductMeta(productId);
  }, [sortStatus, productId]);

  return (
    <Container>
      <h3 ref={reviewsRef}>Ratings & Reviews</h3>
      <Row>
        <Col xs={4}>
          {(productReviews && productMeta) ? (
            <ReviewsBreakdown
              productMeta={productMeta}
              setProductRating={setProductRating}
              selectedRatings={selectedRatings}
              setSelectedRatings={setSelectedRatings}
              setRatingsLength={setRatingsLength}
            />
          ) : null}
        </Col>
        <Col>
          {(productReviews && productMeta) ? (
            <ReviewsList
              productReviews={productReviews}
              characteristics={productMeta.characteristics}
              sortStatus={sortStatus}
              handleSortChange={handleSortChange}
              renderToggle={renderToggle}
              setRenderToggle={setRenderToggle}
              selectedRatings={selectedRatings}
              ratingsLength={ratingsLength}
              productName={productName}
              productId={productId}
            />
          ) : null}
        </Col>
      </Row>
    </Container>
  );
};

Reviews.propTypes = {
  productId: PropTypes.number.isRequired,
  setProductRating: PropTypes.func.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  reviewsRef: PropTypes.object,
  productName: PropTypes.string.isRequired,
};

Reviews.defaultProps = {
  reviewsRef: {},
};

export default Reviews;
