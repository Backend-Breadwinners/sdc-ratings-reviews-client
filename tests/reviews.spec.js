/*eslint-disable*/
import { configure, shallow, mount } from 'enzyme';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
configure({ adapter: new Adapter() });

import axios from 'axios';
import moxios from 'moxios';

import App from '../pages/index.jsx';
import Reviews from '../components/Reviews-Jim/Reviews';
import ReviewsList from '../components/Reviews-Jim/components/ReviewsList';
import Review from '../components/Reviews-Jim/components/Review';
import ReviewsBreakdown from '../components/Reviews-Jim/components/ReviewsBreakdown';

const productReviews = [
  {
      "review_id": 210512,
      "rating": 3,
      "summary": "Aut repellendus vel.",
      "recommend": true,
      "response": "\"Ullam et in.\"",
      "body": "Esse qui et nesciunt. Rerum dicta autem placeat. Eligendi amet reiciendis rerum voluptate et quas maxime consectetur.",
      "date": "2020-09-09T00:00:00.000Z",
      "reviewer_name": "Imelda24",
      "helpfulness": 23,
      "photos": [
          {
              "id": 373584,
              "url": "https://images.unsplash.com/photo-1519241978701-4302ab53de1b?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1000&q=80"
          },
          {
              "id": 373585,
              "url": "https://images.unsplash.com/photo-1532244769164-ff64ddeefa45?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
          }
      ]
  },
  {
      "review_id": 210513,
      "rating": 3,
      "summary": "Suscipit iste minus aperiam facere.",
      "recommend": true,
      "response": null,
      "body": "Est veniam ut distinctio sint vel qui est. Numquam impedit sed blanditiis molestias voluptas officiis ut. Voluptatem consequatur porro ut repellendus tempore adipisci dignissimos. Nihil voluptatibus inventore cum. Velit sit veritatis et est tempore reiciendis autem harum quis.",
      "date": "2020-10-26T00:00:00.000Z",
      "reviewer_name": "Bailee5",
      "helpfulness": 18,
      "photos": [
          {
              "id": 373583,
              "url": "https://images.unsplash.com/photo-1542702942-161ceb2e3d91?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2850&q=80"
          }
      ]
  },
  {
      "review_id": 210511,
      "rating": 4,
      "summary": "Cum asperiores rerum autem voluptas.",
      "recommend": true,
      "response": "\"Culpa vero totam esse.\"",
      "body": "Enim illum ea. Id quo distinctio maiores ut. Porro esse eos et omnis qui architecto magnam. Ut autem autem odio optio sint at et.",
      "date": "2020-12-20T00:00:00.000Z",
      "reviewer_name": "Albert37",
      "helpfulness": 6,
      "photos": [
          {
              "id": 373586,
              "url": "https://images.unsplash.com/photo-1519241978701-4302ab53de1b?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1000&q=80"
          }
      ]
  },
  {
      "review_id": 210510,
      "rating": 3,
      "summary": "Et fugit quia voluptatibus ipsa soluta autem sed.",
      "recommend": true,
      "response": null,
      "body": "Vero repellat saepe laudantium voluptatem voluptatem repellendus et illum doloribus. Possimus eos et cumque consequatur quod atque sed repellat. Sed architecto et aspernatur nisi officiis. Sunt id dolorem eum occaecati. Ea quos ab qui voluptas voluptatum temporibus natus laboriosam. Ipsum voluptas corporis.",
      "date": "2021-02-07T00:00:00.000Z",
      "reviewer_name": "Arely_Olson",
      "helpfulness": 4,
      "photos": [
          {
              "id": 373587,
              "url": "https://images.unsplash.com/photo-1554735490-80893c93b06f?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
          },
          {
              "id": 373588,
              "url": "https://images.unsplash.com/photo-1558422504-3d17542c1799?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
          }
      ]
  },
  {
      "review_id": 210508,
      "rating": 5,
      "summary": "Quis explicabo quis non ad architecto est.",
      "recommend": true,
      "response": null,
      "body": "Dolores ad qui rerum. Hic maxime voluptas unde alias repellat vel quis iusto aliquid. Rem distinctio voluptatem quasi maxime quo dolores quos iure. Nesciunt labore incidunt numquam molestiae vitae.",
      "date": "2020-09-01T00:00:00.000Z",
      "reviewer_name": "Gideon57",
      "helpfulness": 1,
      "photos": [
          {
              "id": 373590,
              "url": "https://images.unsplash.com/photo-1485646979142-d4abb57a876f?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2089&q=80"
          }
      ]
  }
];
const productMeta = {
  "product_id": "18201",
  "ratings": {
      "2": "5",
      "3": "5",
      "4": "1",
      "5": "1"
  },
  "recommended": {
      "true": "12"
  },
  "characteristics": {
      "Fit": {
          "id": 61056,
          "value": "3.6666666666666667"
      },
      "Length": {
          "id": 61057,
          "value": "3.1666666666666667"
      },
      "Comfort": {
          "id": 61058,
          "value": "2.3333333333333333"
      },
      "Quality": {
          "id": 61059,
          "value": "3.0833333333333333"
      }
  }
};

// USE beforeEach() TO START EACH TEST WITH THE SAME CODE (LIKE A WRAPPER)

xdescribe('Reviews Component', () => {

  describe('Jest tests that test the Jest!', () => {
    it('should expect true to equal true', () => {
      expect(true).toEqual(true);
    });
  })

  describe('Parent Component', () => {
    const dummyProps = {
      productId: 0,
      setProductRating: () => {},
      reviewsRef: {},
      productName: 'string',
    }

    it('should render correctly with no props', () => {
      const wrapper = shallow(<Reviews {...dummyProps} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the widget title', () => {
      const wrapper = shallow(<Reviews {...dummyProps} />);
      expect(wrapper.find('.reviews-title').text()).toContain('Ratings & Reviews');
    });
  })

  describe('Reviews Breakdown', () => {

  })

  describe('Reviews List', () => {
    const dummyProps = {
      productReviews: [{
        rating: 0,
        reviewer_name: 'string',
        date: 'string',
        summary: 'string',
        body: 'string',
        photos: [{
          id: 0,
          url: 'string',
        }],
        recommended: true,
        response: 'string',
        helpfulness: 0,
      }],
      sortStatus: 'relevant',
      handleSortChange: () => {},
      renderToggle: true,
      setRenderToggle: () => {},
      selectedRatings: [0],
      ratingsLength: 0,
      characteristics: {
          // id: 0,
          // value: 'string',
      },
      productName: 'string',
      productId: 0,
    }

    it('should have a button that adds two reviews', () => {
      const wrapper = shallow(<ReviewsList {...dummyProps} />);
      expect(wrapper.find('#more-reviews-btn').text()).toBe('More Reviews');
    });

    it('should use relevant as a default sort status', () => {
      const wrapper = shallow(<ReviewsList {...dummyProps} />);
      // force it to call use state
      // find the select inside reviews list -- id or classname
      expect(wrapper.find('select').props().value).toBe('relevant');
    });
  })
});

// console.log(wrapper.debug());