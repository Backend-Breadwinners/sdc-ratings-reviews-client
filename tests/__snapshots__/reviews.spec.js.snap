// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews Component Parent Component should render correctly with no props 1`] = `ShallowWrapper {}`;
